pages/learn/develop/container-contracts.md

```diff
@@ -0,0 +1,112 @@
+
+**Note** If `name`, `type`, or `slug` are omitted, the build will fail.
+
+Multiple requirements may be added to a single contract. The following example requires that the Supervisor version is greater than or equal to 10.6.13, and the L4T version is equal to 32.2.
+
+```yaml
+type: "sw.container"
+slug: "enforce-supervisor-and-l4t"
+name: "Enforce supervisor and l4t requirements"
+requires:
+  - type: "sw.supervisor"
+    version: ">=10.6.17"
+  - type: "sw.l4t"
+    version: "32.2"
+```
+
+**Note** Any type other than `sw.supervisor` or `sw.l4t` will result in the contract being invalid.
+
+When deploying a release, if the contract requirements are not met, the release will fail, and all devices in the application will remain on their current release. This default behavior can be overridden by including an [optional label](#optional-containers) in the application docker-compose.yml file. This label specifies which services with unmet requirements will be ignored, with the release being deployed to all other devices in the application. If there are any existing running containers for services with unmet requirements, they will not be destroyed and continue running the previous release.
+
+Should a contract fail, the dashboard logs will contain detail about the failing contract:
+
+```shell
+Could not move to new release: Some releases were rejected due to having unmet requirements:
+Contracts: Services with unmet requirements: first-service
+```
+
+**Note** For application fleets that contain a mixture of Supervisor and L4T versions, contracts may be used to ensure only compatible devices run specific services as releases will only be deployed to those meeting the contract requirements.
+
+## Optional containers
+
+By default, when a service contract fails, none of the services are deployed to the device. However, in a multi-container application, it is possible to ignore those services that fail the contract requirements with the other services being deployed as normal. To do so, we make use of the `io.balena.features.optional: 1` [Supervisor label][labels] to indicate which services should be considered optional.
+
+In the `docker-compose.yml` file of the application, add the `io.balena.features.optional: 1` to the labels list for each service you wish to mark as optional. In the following example, even if the `first-service` service contract fails, the `second-service` service will still be deployed (assuming it doesn’t have any failing contracts of its own).
+
+```Dockerfile
+version: '2'
+services:
+  first-service:
+    build: ./first-service
+    labels:
+      - io.balena.features.optional: 1
+  second-service:
+    build: ./second-service
+```
+
+**Note** When using optional containers, if there are any existing running services of the failed contract, they will be killed and not continue running the old version.
```

Comment:
Does the optional service get killed, and the optional container left at the old version; or does the optional service get killed, with the optional container being updated but not started?

Reply:
Service is synonymous to container so the service (container) is killed and so there is nothing to run. You noted earlier the term "service container" that I introduced which I have since tidied to make this more clear.

Reply:
Yeah, reads much nicer now IMHO.